src/metadata/metadata.rs

```diff
@@ -0,0 +1,207 @@
+            "Editor" => {
+                let val = parse::parse_editor(val);
+                self.editors.join(Some(val));
+            }
+            "Group" => {
+                let val = val.clone();
+                self.group.join(Some(val));
+            }
+            "Level" => {
+                let val = parse::parse_level(val);
+                self.level.join(Some(val));
+            }
+            "Shortname" => {
+                let val = val.clone();
+                self.shortname.join(Some(val))
+            }
+            "Status" => {
+                let val = val.clone();
+                self.raw_status.join(Some(val));
+            }
+            "Title" => {
+                let val = val.clone();
+                self.title.join(Some(val));
+            }
+            _ => die!("Unknown metadata key \"{}\".", key; line_num),
+        }
+
+        self.has_metadata = true;
+    }
+
+    pub fn fill_macros(&self, doc: &mut Spec) {
+        let macros = &mut doc.macros;
+
+        if let Some(date) = self.date.as_ref() {
+            macros.insert(
+                "date",
+                date.format(&format!("{} %B %Y", date.day())).to_string(),
+            );
+            macros.insert("isodate", date.to_string());
+        }
+        if let Some(level) = self.level.as_ref() {
+            macros.insert("level", level.clone());
+        }
+        if let Some(shortname) = self.shortname.as_ref() {
+            macros.insert("shortname", shortname.clone());
+        }
+        if let Some(raw_status) = self.raw_status.as_ref() {
+            macros.insert(
+                "longstatus",
+                SHORT_TO_LONG_STATUS
+                    .get(raw_status.as_str())
+                    .unwrap()
+                    .to_string(),
+            );
+        }
+        if let Some(title) = self.title.as_ref() {
+            macros.insert("title", title.clone());
+            macros.insert("spectitle", title.clone());
+        }
+    }
+
+    pub fn compute_implicit_metadata(&mut self) {
+        if self.canonical_url.is_none() || self.canonical_url.as_ref().unwrap() == "ED" {
+            self.canonical_url = self.ed.clone();
+        }
+    }
+
+    pub fn validate(&self) {
+        if !self.has_metadata {
+            die!("The document requires at least one <pre class=metadata> block.");
+        }
+    }
+}
+
+pub fn parse_metadata(lines: &Vec<Line>) -> (MetadataManager, Vec<Line>) {
+    let mut mm = MetadataManager::new();
+    let mut new_lines: Vec<Line> = Vec::new();
+    let mut in_metadata = false;
+    let mut last_key: Option<String> = None;
+
+    let title_reg = Regex::new(r"\s*<h1[^>]*>(.*?)</h1>").unwrap();
+    let begin_tag_reg = Regex::new(r"<(pre|xmp) [^>]*class=[^>]*metadata[^>]*>").unwrap();
+    let mut end_tag_reg: Option<Regex> = None;
+    let pair_reg = Regex::new(r"([^:]+):\s*(.*)").unwrap();
+
+    for line in lines {
+        if !in_metadata && begin_tag_reg.is_match(&line.text) {
```

Comment:
This parsing of html using regexes is interesting... Is this what bikeshed does?

Fine if so, just a bit sketchy.

Reply:
Yes, that's exactly what bikeshed does. Maybe there's a better way.

Reply:
Maybe file an issue and reference it from here like `// TODO(#<issue-number>): Figure out if we can get rid of this html-parsing-with-regexes.`